pytorch_toolkit/ote/ote/modules/arg_converters/mmaction.py

```diff
@@ -14,52 +14,66 @@
+            })
+        return cur_map
+
+    def train_update_args_map(self):
+        cur_map = self._train_compression_base_args_map_with_resume_load()
+        cur_map.update({
+            'base_learning_rate': 'optimizer.lr',
+            'epochs': 'total_epochs',
+            })
+        return cur_map
+
+    def test_update_args_map(self):
+        return {
+                'test_ann_files': 'data.test.ann_file',
+                'test_data_roots': 'root_dir',
+               }
+
+    def compress_update_args_map(self):
+        return self._train_compression_base_args_map_with_resume_load()
+
+    def train_to_compress_update_args_map(self):
+        return self._train_compression_base_args_map()
+
+    def train_out_args_map(self):
+        return super().train_out_args_map()
+
+    def compress_out_args_map(self):
+        return super().compress_out_args_map()
+
+    def test_out_args_map(self):
+        return super().test_out_args_map()
+
+    def get_extra_train_args(self, args):
+        return {}
+
+    def get_extra_test_args(self, args):
+        return {}
 
 @ARG_CONVERTERS.register_module()
 class MMActionArgsConverter(BaseArgConverter):
-    # NB: compress_update_args_map is the same as train_update_args_map,
-    #     but without base_learning_rate and epochs
-    # TODO(LeonidBeynenson): replace the dicts by a function that returns dicts to avoid copying of code
-    compress_update_args_map = {
-        'train_ann_files': 'data.train.ann_file',
-        'train_data_roots': 'root_dir',
-        'val_ann_files': 'data.val.ann_file',
-        'val_data_roots': 'root_dir',
-        'resume_from': 'resume_from',
-        'load_weights': 'load_from',
-        'save_checkpoints_to': 'work_dir',
-        'batch_size': 'data.videos_per_gpu',
-    }
-    train_update_args_map = {
-        'train_ann_files': 'data.train.ann_file',
-        'train_data_roots': 'root_dir',
-        'val_ann_files': 'data.val.ann_file',
-        'val_data_roots': 'root_dir',
-        'resume_from': 'resume_from',
-        'load_weights': 'load_from',
-        'save_checkpoints_to': 'work_dir',
-        'batch_size': 'data.videos_per_gpu',
-        'base_learning_rate': 'optimizer.lr',
-        'epochs': 'total_epochs',
-    }
-    train_to_compress_update_args_map = {
-        'train_ann_files': 'data.train.ann_file',
-        'train_data_roots': 'root_dir',
-        'val_ann_files': 'data.val.ann_file',
-        'val_data_roots': 'root_dir',
-# the only difference w.r.t compress_update_args_map
-#        'resume_from': 'resume_from',
-#        'load_weights': 'load_from',
-        'save_checkpoints_to': 'work_dir',
-        'batch_size': 'data.videos_per_gpu',
-    }
-    test_update_args_map = {
-        'test_ann_files': 'data.test.ann_file',
-        'test_data_roots': 'root_dir',
-    }
-
     def __init__(self):
-        super(MMActionArgsConverter, self).__init__()
+        super().__init__(MMActionArgConverterMap())
```

Comment:
2 @evgeny-izutov (CC @Ilya-Krylov , @druzhkov-paul ):   
Sorry, but I insist that it is **not** a good decision to merge `XXXArgsConverterMap` with `XXXArgsCoverter`

The main reason is that when we make inheritance, we do not know which inherited methods or fields are changed in derived classes and which not. So, we cannot make decomposition.

In my approach we do not change any methods/fields in the classes derived from `BaseArgConverter` : the only purpose of the classes are to be registered in the proper `Registry`
Instead of changing methods in classes derived from `BaseArgConverter` now we have a special base abstract class `ArgConverterMaps` that contains all the methods that should be implemented for a concrete arg converter, all of the methods are marked as `@abstractmethod`, and all the `XXXArgConverterMap` classes implement them. 

Moreover, almost all `XXXArgConverterMap` classes are derived just from `ArgConverterMaps` , so we have height of class tree = 1 almost everywhere
(the only exception is `MMDetectionWiderArgConverterMap` that is derived from `MMDetectionArgConverterMap` and changes one method only --  we can consider this)

As the result we have decomposition: 
* the classes `XXXArgConverterMaps` implement the methods that are individual for each arg converter, 
* `BaseArgConverter` just combines calls of several functions from some class `XXXArgConverterMaps` -- they are  are external w.r.t. this class, since they are passed to `BaseArgConverter` in the constructor 
* `XXXArgsConverter` just connects `BaseArgConverter` with `XXXArgConverterMaps`: it implements one constructor and is registered in the correponding `Registry`

Note that `XXXArgsConverter`  classes are required since  `mmcv.Registry` is used -- otherwise a factory function would be sufficient.

Reply:
Yes, I can read the code too - so there is no need at describing the functionality again in words (so many words)... Moreover, I'm former adept of Design Patterns but the main question - do we actually need at code complicating for the end user here? I mean the code of each `XXXArgsConverter` is the same - so we can support registry of `XXXArgConverterMap` instead and move the creation of `BaseArgConverter` into `XXX_build` function. It will allow the end user to redefine `XXXArgConverterMap` class only - if I'm right.

Reply:
I'm really sorry, I thought that misunderstanding was happened.

Yes, the code of `XXXArgsConverter` is the same -- but it is 2-3 lines of code, it is just a connector.

Just to be on the safe side: do you propose to make `build_arg_converter` a special function that does not just calls `build` with a Registry, but makes `build` with a registry to create a  "Map" and then instantiates `ArgsConverter` (ex `BaseArgsConverter`) with the created map?

I'm generally ok with this change, but it breaks the uniformity of the code: at the moment to instantiate each "operation" (arg parser, arg converter, etc) one `build...` function is required (that just makes `build` in the corresponding registry)

In my opinion uniformity of the code is a bit more important.

Reply:
If the only purpose to create a beautiful code - I agree, the current structure is perfect! But from other hand, the user who wants to add own an repo into the ote should implement not only task-related classes but some architecture-related code too. It's unfriendly way, IMHO. And yes, the proposed changes is about hiding architecture-related activity from the potential user of model_templates (not ote).